dowhy/gcm/distribution_change_robust.py

```diff
@@ -0,0 +1,915 @@
+    :param crop: float, all predicted probabilities from the classifier will be cropped below at this lower bound,
+            and above at 1-crop.
+
+    :param shapley_config: Configuration for the Shapley estimator.
+
+    :return: A dictionary containing contribution of each upstream node.
+    """
+
+    causal_graph = causal_model.graph
+
+    # Eliminate nodes that are not ancestry of the outcome:
+    causal_graph = nx.DiGraph(node_connected_subgraph_view(causal_model.graph, target_node))
+
+    # Sort by causal ancestry:
+    sorted_var_names = list(nx.lexicographical_topological_sort(causal_graph))
+    X_var_names = sorted_var_names[:-1]
+
+    n0, n1 = old_data.shape[0], new_data.shape[0]
+
+    X0 = old_data[X_var_names].values
+    X1 = new_data[X_var_names].values
+    y0 = old_data[target_node].values.flatten()
+    y1 = new_data[target_node].values.flatten()
+    X = np.concatenate((X0, X1))
+    y = np.concatenate((y0, y1))
+    T = np.concatenate((np.zeros(n0), np.ones(n1)))
+
+    if not xfit:
+        if sample_weight is None:
+            X_train, X_eval, y_train, y_eval, T_train, T_eval = train_test_split(
+                X, y, T, train_size=train_size, stratify=T, random_state=split_random_state
+            )
+            X_calib, T_calib = None, None
+
+            if calibrator is not None:
+                if calib_size == 0.0:
+                    raise ValueError(
+                        "For calibration, calib_size should be either positive and smaller than the number of samples or a float in the (0, 1) range."
+                    )
+
+                X_calib, X_train, _, y_train, T_calib, T_train = train_test_split(
+                    X_train, y_train, T_train, train_size=calib_size, stratify=T_train, random_state=split_random_state
+                )
+
+            w_eval, w_train, w_calib = None, None, None
+
+        else:
+            w = np.concatenate((old_data[sample_weight].values.flatten(), new_data[sample_weight].values.flatten()))
+
+            X_train, X_eval, y_train, y_eval, T_train, T_eval, w_train, w_eval = train_test_split(
+                X, y, T, w, train_size=train_size, stratify=T, random_state=split_random_state
+            )
+            X_calib, T_calib, w_calib = None, None, None
+
+            if calibrator is not None:
+                if calib_size == 0.0:
+                    raise ValueError(
+                        "For calibration, calib_size should be either positive and smaller than the number of samples or a float in the (0, 1) range."
+                    )
+
+                X_calib, X_train, _, y_train, T_calib, T_train, w_calib, w_train = train_test_split(
+                    X_train,
+                    y_train,
+                    T_train,
+                    w_train,
+                    train_size=calib_size,
+                    stratify=T_train,
+                    random_state=split_random_state,
+                )
+
+        def set_func(C):
```

Comment:
Can the set function be defined outside of the `if xfit` part here to avoid the code duplication ? Seems very similar code (might missed a small detail). In that case, just define the arguments in the if-else parts and then define the set function afterwards.

Reply:
I've modified this to avoid code duplication.